### PR TITLE
Introduce temporary if necessary to inline `_apply`

### LINKED ISSUE #21065: Distressing performance unless promotion is manual
Given the impact of the splatting penalty in other areas, I've long wondered why [these definitions](https://github.com/JuliaLang/julia/blob/c3f2f4ccca5da3465f278a0c6beb0bdf3d4b3bf1/base/promotion.jl#L246-L249) aren't problematic. It turns out that, in some circumstances, they are:
```julia
julia> using FixedPointNumbers

julia> x = 0.2f0
0.2f0

julia> y = N0f8(0.2)
0.2N0f8

julia> @code_llvm x*y

define float @"julia_*_68175"(float, %Normed*) #0 !dbg !5 {
top:
  %ptls_i8 = call i8* asm "movq %fs:0, $0;\0Aaddq $$-10896, $0", "=r,~{dirflag},~{fpsr},~{flags}"() #3
  %ptls = bitcast i8* %ptls_i8 to i8****
  %2 = alloca [4 x i8**], align 8
  %.sub = getelementptr inbounds [4 x i8**], [4 x i8**]* %2, i64 0, i64 0
  %3 = getelementptr [4 x i8**], [4 x i8**]* %2, i64 0, i64 2
  %4 = bitcast i8*** %3 to i8*
  call void @llvm.memset.p0i8.i32(i8* %4, i8 0, i32 16, i32 8, i1 false)
  %5 = bitcast [4 x i8**]* %2 to i64*
  store i64 4, i64* %5, align 8
  %6 = getelementptr [4 x i8**], [4 x i8**]* %2, i64 0, i64 1
  %7 = bitcast i8* %ptls_i8 to i64*
  %8 = load i64, i64* %7, align 8
  %9 = bitcast i8*** %6 to i64*
  store i64 %8, i64* %9, align 8
  store i8*** %.sub, i8**** %ptls, align 8
  %10 = getelementptr [4 x i8**], [4 x i8**]* %2, i64 0, i64 3
  %11 = call [2 x float] @julia_promote_68176(float %0, %Normed* %1)
  %.elt = extractvalue [2 x float] %11, 0
  %.elt2 = extractvalue [2 x float] %11, 1
  store i8** inttoptr (i64 139997130231784 to i8**), i8*** %3, align 8
  %12 = call i8** @jl_gc_pool_alloc(i8* %ptls_i8, i32 1440, i32 16)
  %13 = getelementptr i8*, i8** %12, i64 -1
  %14 = bitcast i8** %13 to i8***
  store i8** inttoptr (i64 139997180171536 to i8**), i8*** %14, align 8
  %15 = bitcast i8** %12 to float*
  store float %.elt, float* %15, align 8
  %.sroa_raw_cast5 = bitcast i8** %12 to i8*
  %.sroa_raw_idx6 = getelementptr inbounds i8, i8* %.sroa_raw_cast5, i64 4
  %16 = bitcast i8* %.sroa_raw_idx6 to float*
  store float %.elt2, float* %16, align 4
  store i8** %12, i8*** %10, align 8
  %17 = call i8** @jl_f__apply(i8** null, i8*** %3, i32 2)
  %18 = bitcast i8** %17 to float*
  %19 = load float, float* %18, align 16
  %20 = load i64, i64* %9, align 8
  store i64 %20, i64* %7, align 8
  ret float %19
}

julia> function myprod1(x, y)
           xp, yp = promote(x, y)
           xp*yp
       end
myprod1 (generic function with 1 method)

julia> function myprod2(x, y)
           T = promote_type(typeof(x), typeof(y))
           convert(T, x) * convert(T, y)
       end
myprod2 (generic function with 1 method)

julia> @code_llvm myprod1(x, y)

define float @julia_myprod1_68215(float, %Normed*) #0 !dbg !5 {
top:
  %2 = call [2 x float] @julia_promote_68176(float %0, %Normed* %1)
  %.elt = extractvalue [2 x float] %2, 0
  %.elt2 = extractvalue [2 x float] %2, 1
  %3 = fmul float %.elt, %.elt2
  ret float %3
}

julia> @code_llvm myprod2(x, y)

define float @julia_myprod2_68216(float, %Normed*) #0 !dbg !5 {
top:
  %2 = getelementptr inbounds %Normed, %Normed* %1, i64 0, i32 0
  %3 = load i8, i8* %2, align 1
  %4 = uitofp i8 %3 to float
  %5 = fmul fast float %4, 0x3F70101020000000
  %6 = fmul float %5, %0
  ret float %6
}

julia> using BenchmarkTools

julia> @benchmark $x*$y seconds=1
BenchmarkTools.Trial: 
  memory estimate:  64 bytes
  allocs estimate:  4
  --------------
  minimum time:     73.504 ns (0.00% GC)
  median time:      74.888 ns (0.00% GC)
  mean time:        79.331 ns (3.29% GC)
  maximum time:     1.387 μs (87.63% GC)
  --------------
  samples:          10000
  evals/sample:     972
  time tolerance:   5.00%
  memory tolerance: 1.00%

julia> @benchmark myprod1($x, $y) seconds=1
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     5.037 ns (0.00% GC)
  median time:      5.041 ns (0.00% GC)
  mean time:        5.101 ns (0.00% GC)
  maximum time:     18.131 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%

julia> @benchmark myprod2($x, $y) seconds=1
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     3.703 ns (0.00% GC)
  median time:      3.706 ns (0.00% GC)
  mean time:        3.738 ns (0.00% GC)
  maximum time:     14.419 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

Note that this is in stark contrast to the case with "conventional" types:
```julia
julia> z = 3
3

julia> @which x*z
*(x::Number, y::Number) in Base at promotion.jl:247

julia> @which x*y
*(x::Number, y::Number) in Base at promotion.jl:247

julia> @benchmark $x*$z seconds=1
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     2.025 ns (0.00% GC)
  median time:      2.030 ns (0.00% GC)
  mean time:        2.060 ns (0.00% GC)
  maximum time:     12.631 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%

julia> @code_llvm x*z

define float @"julia_*_68579"(float, i64) #0 !dbg !5 {
top:
  %2 = sitofp i64 %1 to float
  %3 = fmul float %2, %0
  ret float %3
}
```
Before I make the obvious fix, is there a deeper issue to understand here?

Thanks to @Evizero for the observation that led to this investigation. CC @vchuravy, since it seriously affects FPN.

### PR DESCRIPTION
**Pull request by martinholters**
With
```julia
julia> function foo(x,y)
           println("x=$x, y=$y")
           return x, y
       end
foo (generic function with 1 method)

julia> bar(x,y) = +(foo(x,y)...)
bar (generic function with 1 method)
```
master:
```julia
julia> @code_typed bar(1, 2)
CodeInfo(:(begin 
        return (Core._apply)(Main.+, $(Expr(:invoke, MethodInstance for foo(::Int64, ::Int64), :(Main.foo), :(x), :(y))))::Int64
    end))=>Int64
```
this PR:
```julia
julia> @code_typed bar(1, 2)
CodeInfo(:(begin 
        #temp# = $(Expr(:invoke, MethodInstance for foo(::Int64, ::Int64), :(Main.foo), :(x), :(y)))
        return (Base.add_int)((Core.getfield)(#temp#, 1)::Int64, (Core.getfield)(#temp#, 2)::Int64)::Int64
    end))=>Int64
```

Fixes #21065 and makes #21067 (almost?) obsolete.  (The "almost" being a matter of inlining `promote`, `promote_type`, and `convert`, where #21067 might still be advantageous.)

### CODE REVIEW

**Comment by martinholters** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

File: base/inference.jl [hunk 1]
```diff
@@ -4462,10 +4462,18 @@ function inlining_pass(e::Expr, sv::InferenceState, stmts, ins)
...
                 elseif isa(t, DataType) && t.name === Tuple.name && !isvatuple(t) &&
-                        effect_free(aarg, sv.src, sv.mod, true) && length(t.parameters) <= sv.params.MAX_TUPLE_SPLAT
-                    # apply(f,t::(x,y)) => f(t[1],t[2])
+                         length(t.parameters) <= sv.params.MAX_TUPLE_SPLAT
+                    if effect_free(aarg, sv.src, sv.mod, true)
+                        # apply(f,t::(x,y)) => f(t[1],t[2])
+                        tmpv = aarg
+                    else
+                        # apply(f,t::(x,y)) => tmp=t; f(tmp[1],tmp[2])
+                        tmpv = add_slot!(sv.src, t, false)
```
Comment (JeffBezanson) [inline base/inference.jl]: I think this should use `newvar!` and make an SSAValue instead. (That function is badly named; should be e.g. `new_ssavalue`.)

**Comment by martinholters** [comment]
I wasn't sure whether the `effect_free` test is still necessary. Is there any reason the SSAValue should not be introduced unconditionally?

Also, what's a good way to test this PR? Inspect `code_typed(bar, Tuple{Int,Int})[1].first.code` and verify there is no `:(Core._apply)` in there? Do we have similar tests somewhere?

File: base/inference.jl (hunk 1 above)
Comment (martinholters) [inline base/inference.jl]: Updated accordingly.

**Comment by martinholters** [comment]
Added a test, but it does look a bit messy...

**Review by JeffBezanson — APPROVED** [review]

**Comment by JeffBezanson** [comment]
Looks fine to me. There is a test kind of similar to this in test/inline.jl, but we usually just test performance and behavior.

Keeping the `effect_free` test is fine. We could introduce the ssavalue unconditionally and let it get removed by `remove_redundant_temp_vars!`, but it doesn't really matter.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21069#issuecomment-287374695) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/2a199f4bf6688fd4fe1478d210cd2e94655009ad/b4df9c3_vs_b549d2b/report.md). cc @jrevels

**Comment by tkelman** [comment]
would have been better to squash merge this

**Pull request merged**